youtube_dl/extractor/nhk.py

```diff
@@ -3,57 +3,38 @@
-    # Content available only for a limited period of time. Visit
-    # https://www3.nhk.or.jp/nhkworld/en/ondemand/ for working samples.
-    _TESTS = [{
-        # clip
-        'url': 'https://www3.nhk.or.jp/nhkworld/en/ondemand/video/9999011/',
-        'md5': '256a1be14f48d960a7e61e2532d95ec3',
-        'info_dict': {
-            'id': 'a95j5iza',
-            'ext': 'mp4',
-            'title': "Dining with the Chef - Chef Saito's Family recipe: MENCHI-KATSU",
-            'description': 'md5:5aee4a9f9d81c26281862382103b0ea5',
-            'timestamp': 1565965194,
-            'upload_date': '20190816',
-        },
-    }, {
-        'url': 'https://www3.nhk.or.jp/nhkworld/en/ondemand/video/2015173/',
-        'only_matching': True,
-    }, {
-        'url': 'https://www3.nhk.or.jp/nhkworld/en/ondemand/audio/plugin-20190404-1/',
-        'only_matching': True,
-    }, {
-        'url': 'https://www3.nhk.or.jp/nhkworld/fr/ondemand/audio/plugin-20190404-1/',
-        'only_matching': True,
-    }, {
-        'url': 'https://www3.nhk.or.jp/nhkworld/en/ondemand/audio/j_art-20150903-1/',
-        'only_matching': True,
-    }]
-    _API_URL_TEMPLATE = 'https://api.nhk.or.jp/nhkworld/%sod%slist/v7a/episode/%s/%s/all%s.json'
+class NhkBaseInfoExtractor(InfoExtractor):
+    _API_URL_TEMPLATE = 'https://api.nhk.or.jp/nhkworld/%sod%slist/v7a/%s/%s/%s/all%s.json'
 
-    def _real_extract(self, url):
-        lang, m_type, episode_id = re.match(self._VALID_URL, url).groups()
-        if episode_id.isdigit():
-            episode_id = episode_id[:4] + '-' + episode_id[4:]
+    def _get_clean_field(self, episode, key):
+        return episode.get(key + '_clean') or episode.get(key)
 
-        is_video = m_type == 'video'
-        episode = self._download_json(
+    def _list_episodes(self, lang, is_video, is_episode, m_id):
+        return self._download_json(
             self._API_URL_TEMPLATE % (
                 'v' if is_video else 'r',
-                'clip' if episode_id[:4] == '9999' else 'esd',
-                episode_id, lang, '/all' if is_video else ''),
-            episode_id, query={'apikey': 'EJfK8jdS57GqlupFgAfAAwr573q01y6k'})['data']['episodes'][0]
+                'clip' if m_id[:4] == '9999' else 'esd',
+                'episode' if is_episode else 'program',
+                m_id, lang, '/all' if is_video else ''),
+            m_id, query={'apikey': 'EJfK8jdS57GqlupFgAfAAwr573q01y6k'})['data']['episodes']
+
+    def _parse_episode_json(self, episode, lang, is_video):
         title = episode.get('sub_title_clean') or episode['sub_title']
 
-        def get_clean_field(key):
-            return episode.get(key + '_clean') or episode.get(key)
+        pgm_id = episode.get('pgm_id')
+        pgm_no = episode.get('pgm_no')
+
+        if not (pgm_id and pgm_no):
+            raise ExtractorError('Cannot download episode: %s' % episode, expected=True)
```

Comment:
you have to test the code, and see if the result is usefull, thos code with result in error message similar to this:
```
Cannot download episode: {'data': {'episodes': [{'pgm_gr_id': 'dwc', 'image': '/nhkworld/en/ondemand/video/9999011/images/buCOFSNvcN0ctbWvHm25sLdGTOKRUX32w7YLHHTS.jpeg', 'image_l': '/nhkworld/en/ondemand/video/9999011/images/dYFYZODb5cmlf6efJ9Cgb4bWIouBc6M83TEhRMFZ.jpeg', 'image_promo': '', 'voice_lang': 'en', 'caption_langs': [], 'voice_langs': ['en'], 'vod_id': 'lpZXIwaDE6_Z-976CPsFdxyICyWUzlT5', 'movie_lengh': '2:28', 'movie_duration': 148, 'analytics': "[nhkworld]clip;Dining with the Chef_Chef Saito's Family recipe: MENCHI-KATSU;en,001;9999-011-1970;", 'title': 'Dining with the Chef', 'title_clean': 'Dining with the Chef', 'sub_title': "Chef Saito's Family recipe: MENCHI-KATSU", 'sub_title_clean': "Chef Saito's Family recipe: MENCHI-KATSU", 'description': 'Umami juice splashes out from crunchy Panko batter! Menchi-katsu is a mouthwatering deep fried meat ball from Japan. [From "Dining with the Chef"]', 'description_clean': 'Umami juice splashes out from crunchy Panko batter! Menchi-katsu is a mouthwatering deep fried meat ball from Japan. [From "Dining with the Chef"]', 'url': '/nhkworld/en/ondemand/video/9999011/', 'category': [17], 'mostwatch_ranking': 136, 'related_episodes': [], 'tags': ['washoku'], 'chapter_list': [], 'life': 0, 'life_category': [], 'promotion': []}]}}
```
the error message is meant to be readable and usefull to the user, that is the point that i'm try pass.

Reply:
See latest revision. Does that seem more helpful?

Reply:
> Does that seem more helpful? If this error was to actually happen, it would mean that what the API returns has changed beyond what we expect. When the happens, I think all the user can do is report an issue here.

there is a generic option to do this(`--dump-pages`), and would ask to the user to provide this output if needed.

> Is it typical practice to say in error messages "report this error to youtube-dl"? If so I can add that as well.

you can do that, by removing `expected=True` from the options passed to `ExtractorError`,

Reply:
@remitamine, ok, expected is now false. Thanks for teaching me that. Also, I am choosing to leave the JSON in because if the user does report the error, I think it'd be helpful here to know what the API returned in case it is hard to reproduce.

Reply:
Oh I see that --dump-pages will print the API response. I guess that could mean I can remove the JSON from the string as well. Let me know what is considered best practice for youtube-dl and I'll go with that.

Reply:
> expected is now false

no need to type `expected=False`, it's the default.

> Oh I see what you mean about --dump-pages, it will print the API response. I guess that could mean I can remove the JSON from the string as well.

Exactly.

Reply:
😊